encoding/mixed_readers_iterator.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"github.com/m3db/m3db/interfaces/m3db"
+	xtime "github.com/m3db/m3db/x/time"
+)
+
+type activeIterState int
+
+const (
+	activeIterStateNone activeIterState = iota
+	activeIterStateSingle
+	activeIterStateMulti
+)
+
+type mixedReadersIterator struct {
+	singleIter m3db.SingleReaderIterator
+	multiIter  m3db.MultiReaderIterator
+	slicesIter m3db.ReaderSliceOfSlicesIterator
+	state      activeIterState
+	err        error
+	closed     bool
+	pool       m3db.MixedReadersIteratorPool
+}
+
+// NewMixedReadersIterator creates a new mixed readers iterator
+func NewMixedReadersIterator(
+	singleIter m3db.SingleReaderIterator,
+	multiIter m3db.MultiReaderIterator,
+	slicesIter m3db.ReaderSliceOfSlicesIterator,
+	pool m3db.MixedReadersIteratorPool,
+) m3db.MixedReadersIterator {
+	it := &mixedReadersIterator{
+		singleIter: singleIter,
+		multiIter:  multiIter,
+		pool:       pool,
+	}
+	it.Reset(slicesIter)
+	return it
+}
+
+func (it *mixedReadersIterator) Next() bool {
+	if it.err != nil || it.closed {
+		return false
+	}
+	switch it.state {
+	case activeIterStateNone:
+		next := it.slicesIter.Next()
+		if !next {
+			return false
+		}
+
+		readers := it.slicesIter.Current()
```

Comment:
Would it make sense to wrap up this logic inside `ReaderSliceOfSlicesIterator` instead? i.e., instead of returning `[]io.Reader`, `Current` in `ReaderSliceOfSlicesIterator` will return what `Current` in `mixedReadersIterator` returns. This way we don't need the `MixedReadersIterator` interface and its corresponding pool, which could make the interfaces a bit cleaner.